### DISCUSSION TITLE
Question about preventing dokan delete a file.

### DISCUSSION THREAD

**Discussion by Lucasq11** [dokan-dev/dokany#1146] [Q&A]
@zqlustc Per microsoft documentation, cleanup cannot fail. If delete must fail, it must be during createfile or set delete that everything need to put in place for the delete to work during cleanup.

_Originally posted by @Liryna in https://github.com/dokan-dev/dokany/issues/970#issuecomment-774614866_

I want to control the right of deleting a file, I know that cleanup can't be stopped and try to use another fuction: DeleteFile in dokan_operations.

the interface comment says:
 DeleteFile will also be called with DOKAN_FILE_INFO.DeleteOnClose set to \c FALSE to notify the driver when the file is no longer requested to be deleted.
And DeleteFile can return STATUS_ACCESS_DENIED to stop delete.

But I find DeleteFile is not called when I delete a file in a file system created by the memfs.exe, so how can I control the right of delete.

**Comment by Liryna** [comment]
Windows can mark the file for deletion directly at `CreateFile` time with [FILE_DELETE_ON_CLOSE](https://learn.microsoft.com/en-us/windows-hardware/drivers/ifs/delete-on-close) flag. You can return `STATUS_ACCESS_DENIED` at that time to cancel the delete.

`DeleteFile` is used to mark a file for deletion while the handle was already open and without `FILE_DELETE_ON_CLOSE`.

**Reply by Lucasq11** [reply]
nice of you, I'll try it.

**Reply by Lucasq11** [reply]
Thanks, it worked!

**Reply by Lucasq11** [reply]
Sorry for another question. I found that last time I stopped the deletation by setting both return values to STATUS_ACCESS_DENIED  of CreateFile and DeleteFile. If I only set STATUS_ACCESS_DENIED when CreateFile is called when meet FILE_DELETE_ON_CLOSE, the deletion can't be stopped. I printed the log and it shows that DeleteFile is also called after CreatedFile, and it returns 0. I guess this is the reseaon that deletion can't be stopped. But when I set STATUS_ACCESS_DENIED to DeleteFile, I tried some other software, like Adobe Illustrator. It can't save files to my file system, when I disable the return value STATUS_ACCESS_DENIED in DeleteFile, It can save file, but at this time, deletion is not avaliable. Do you know why DeleteFile can cancel a file even CreateFile returns STATUS_ACCESS_DENIED?

**Reply by Liryna** [reply]
Hum Delete file can't be called if createfile failed. I believe what you see is a delete called on another handle open by a createfile without the delete on close flag.

**Reply by Lucasq11** [reply]
Sorry that I don't kown how to enable log. Here are some simple logs printed by myself of CreateFile and DeleteFile in dokanfuse/src. It roughly recored function call stack when I try to delete a file whoes path is \test_notebook1.txt. CreateDisposition and CreateOptions are in hexdecimal. 

Do you mean that DeleteFile is called by calling CreateFile in line 14? And why an empty path is called at line 6?

I also found that in all the CreateFile call in this procedure, CreateOptions has the value 200040 only at line 14. What functions does this value of CreateOption cover? Can I return STATUS_ACCESS_DENIED for CreateFile directly when 200040 is meet?

1. ENTER FuseCreateFile, path: \test_notebook1.txt, disposition: 1, options: 1040, DELETE_ON_CLOSE: 0
2. ENTER impl_fuse_context::create_file, path: \test_notebook1.txt, disposition: 1, options: 1040, DELETE_ON_CLOSE: 0
3. impl_fuse_context::create_file path: \test_notebook1.txt, catch delete on close, call ops_.unlink
4. RETURN impl_fuse_context::create_file path: \test_notebook1.txt, res is: -13, DELETE_ON_CLOSE is: 0
5. RETURN FuseCreateFile, path: \test_notebook1.txt, res: c0000022, DELETE_ON_CLOSE is: 0
6. ENTER FuseCreateFile, path: , disposition: 1, options: 5000, DELETE_ON_CLOSE: 0
7. ENTER impl_fuse_context::create_file, path: , disposition: 1, options: 5000, DELETE_ON_CLOSE: 0
8. RETURN FuseCreateFile, path: , res: c0000034, DELETE_ON_CLOSE is: 0
9. ENTER FuseCreateFile, path: \, disposition: 1, options: 4021, DELETE_ON_CLOSE: 0
10. ENTER FuseCreateFile, path: \, disposition: 1, options: 4021, DELETE_ON_CLOSE: 0
11. ENTER FuseCreateFile, path: \, disposition: 1, options: 4021, DELETE_ON_CLOSE: 0
12. ENTER FuseCreateFile, path: \, disposition: 1, options: 4021, DELETE_ON_CLOSE: 0
13. ENTER FuseCreateFile, path: \, disposition: 1, options: 4021, DELETE_ON_CLOSE: 0
14. ENTER FuseCreateFile, path: \test_notebook1.txt, disposition: 1, options: 200040, DELETE_ON_CLOSE: 0
15. ENTER impl_fuse_context::create_file, path: \test_notebook1.txt, disposition: 1, options: 200040, DELETE_ON_CLOSE: 0
16. RETURN FuseCreateFile, path: \test_notebook1.txt, res: 0, DELETE_ON_CLOSE is: 0
17. ENTER FuseDeleteFile, path: \test_notebook1.txt, DELETE_ON_CLOSE is: 1
18. ENTER impl_fuse_context::delete_file，path=\test_notebook1.txt, DELETE_ON_CLOSE is: 1
19. RETURN impl_fuse_context::delete_file，path=\test_notebook1.txt, res: 0
20. RETURN FuseDeleteFile, path: \test_notebook1.txt, res: 0
